File: apple/LookoutMarine/Charts/ChartGallery.swift
//  ChartGallery.swift: the charts to draw, as tiles.
//
//  One chart draws at a time. Lookout's own chart is built from the installed
//  sets; a link is a publisher's style drawn instead of it. A row of tiles
//  rather than a list, so two styles with similar names are told apart by
//  looking.
//
//  The core draws each tile's picture: the chart on screen as the engine draws
//  it, and one publisher tile for any other raster style. A chart with no
//  picture shows its kind and its url.

import SwiftUI

struct ChartGallery: View {
    var model: AppModel
    /// What the last tile does. The parent owns it, because adding a chart by
    /// link and adding one from a file are the same decision to a mariner and
    /// the parent is where that sheet lives.
    let onAdd: () -> Void

    private var links: ChartLinksModel { model.chartLinks }
    /// One tile of each linked chart, at the water the mariner is on.
    @State private var previews = ChartPreviews()
    @Environment(\.displayScale) private var displayScale

    var body: some View {
        ScrollView(.horizontal) {
            HStack(alignment: .top, spacing: Metrics.gap) {
                ChartCard(style: .tile,
                          name: "Lookout chart",
                          detail: lookoutDetail,
                          active: links.active == nil,
                          picture: previews.images[""] ?? Image("WelcomeChart")) {
                    links.select(nil)
                }
                ForEach(links.list) { link in
                    ChartCard(style: .tile,
                              name: link.name,
                              detail: link.url,
                              active: links.active == link.url,
                              picture: previews.images[link.url],
                              onRefresh: { links.refresh(link.url) },
                              onRemove: { links.remove(link.url) }) {
                        links.select(link.url)
                    }
                }
                AddChartTile(add: onAdd)
            }
            .padding(.vertical, 2)
            .padding(.horizontal, 12)
        }
        .onAppear {
            previews.bind(to: model.controller)
            askPictures()
        }
        // A picture finishing, a pick and a new link all change the revision.
        .onChange(of: links.revision) { _, _ in askPictures() }
        .onDisappear { previews.stop() }
        .scrollIndicators(.automatic)
        // The row starts at its first tile. Without this the form can hand the
        // scroll view an offset and the active tile is cut off at the left.
        .defaultScrollAnchor(.leading)
    }

    /// Every tile's picture, at the water the mariner is on. The zoom is low
    /// enough that one publisher tile holds a recognisable stretch of coast.
    private func askPictures() {
        guard let at = model.controller?.viewCenter() else { return }
        previews.ask([""] + links.list.map(\.url), kind: .tile, lon: at.lon, lat: at.lat,
                     zoom: 9,
                     width: Int((Metrics.tile * displayScale).rounded()),
                     height: Int((Metrics.art * displayScale).rounded()))
    }

    /// What Lookout's own chart is built from.
    private var lookoutDetail: String {
        let cells = model.charts.sets.filter(\.on).reduce(0) { $0 + $1.cells.count }
        if cells == 0 { return "From your chart sets" }
        return "From your chart sets · \(cells) cells"
    }

    enum Metrics {
        #if os(macOS)
        static let tile: CGFloat = 250
        static let art: CGFloat = 132
        static let add: CGFloat = 176
        static let gap: CGFloat = 10
        #else
        static let tile: CGFloat = 196
        static let art: CGFloat = 112
        static let add: CGFloat = 150
        static let gap: CGFloat = 10
        #endif
    }
}


/// One chart: a picture of it where there is one, its name, and where it comes
/// from.
///
/// `.tile` is the gallery's fixed-width tile, with an ACTIVE badge and a menu.
/// `.shelf` is the first-run card: a grid cell with a tick and a remove
/// button, whose url wraps to two lines. A style link holds the publisher,
/// the style and often a key, and those are the parts a middle ellipsis
/// removes first.
struct ChartCard: View {
    enum Style { case tile, shelf }

    let style: Style
    let name: String
    let detail: String
    let active: Bool
    /// One tile of this chart, once it has been fetched.
    var picture: Image? = nil
    /// True while this chart's picture is drawn.
    var drawing = false
    var onRefresh: (() -> Void)? = nil
    var onRemove: (() -> Void)? = nil
    let select: () -> Void

    private var shelf: Bool { style == .shelf }

    var body: some View {
        Button(action: select) {
            VStack(alignment: .leading, spacing: 0) {
                art
                caption
                    .padding(.horizontal, 11)
                    .padding(.top, 9)
                    .padding(.bottom, 11)
            }
            .frame(width: shelf ? nil : ChartGallery.Metrics.tile)
            .modifier(CardSkin(picked: active, radius: shelf ? 12 : 11))
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(active ? [.isButton, .isSelected] : .isButton)
        .accessibilityIdentifier("chart-tile-\(name)")
    }

    private var caption: some View {
        VStack(alignment: .leading, spacing: 3) {
            HStack(spacing: 7) {
                if shelf {
                    Image(systemName: active ? "checkmark.circle.fill" : "circle")
                        .font(.system(size: 14))
                        .foregroundStyle(active ? Chrome.accent : Chrome.ink.opacity(0.30))
                }
                Text(name)
                    .font(.system(size: shelf ? 14 : 13, weight: .semibold))
                    .foregroundStyle(Chrome.ink)
                    .lineLimit(1)
                Spacer(minLength: 0)
                if shelf, let onRemove {
                    Button(action: onRemove) {
                        Image(systemName: "minus.circle")
                    }
                    .buttonStyle(.plain)
                    .foregroundStyle(Chrome.muted)
                    .accessibilityLabel("Remove \(name)")
                }
            }
            Text(detail)
                .font(shelf ? .system(size: 10.5, design: .monospaced) : .system(size: 11))
                .foregroundStyle(Chrome.muted)
                .lineLimit(shelf ? 2 : 1)
                .truncationMode(.middle)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    @ViewBuilder private var art: some View {
        switch style {
        case .tile:
            chartPicture
                .frame(width: ChartGallery.Metrics.tile, height: ChartGallery.Metrics.art)
                .clipped()
                // Over the clipped tile, and not inside it. A picture that fills
                // by covering is wider than the tile, so a badge aligned inside
                // it started left of the tile's own edge and lost its first
                // letter.
                .overlay(alignment: .topLeading) { badge }
                .overlay(alignment: .topTrailing) { menu }
                .overlay(alignment: .bottom) { rule }
        case .shelf:
            chartPicture
                .frame(maxWidth: .infinity)
                // Nearer square than a strip: a chart style is told apart by
                // its water and its marks, and both want height.
                .aspectRatio(4.0 / 3.0, contentMode: .fill)
                .frame(maxHeight: 210)
                .clipped()
                .overlay(alignment: .bottom) { rule }
        }
    }

    private var rule: some View {
        Rectangle().fill(Chrome.edge.opacity(0.5)).frame(height: 1)
    }

    /// The picture at the top of the card, or the room it will take. A style
    /// with vector tiles has no publisher tile to show.
    @ViewBuilder private var chartPicture: some View {
        if let picture {
            picture
                .resizable()
                .aspectRatio(contentMode: .fill)
        } else {
            Chrome.panel
                .overlay {
                    if drawing {
                        VStack(spacing: 7) {
                            ProgressView().controlSize(.small)
                            Text("Drawing this chart…")
                                .font(.system(size: 11.5))
                                .foregroundStyle(Chrome.muted)
                        }
                    } else {
                        Image(systemName: "globe.americas")
                            .font(.system(size: shelf ? 20 : 26, weight: .light))
                            .foregroundStyle(Chrome.accent.opacity(shelf ? 0.5 : 0.55))
                    }
                }
        }
    }

    @ViewBuilder private var badge: some View {
        if active {
            Text("ACTIVE")
                .font(.system(size: 10, weight: .bold))
                .kerning(0.3)
                .foregroundStyle(.white)
                .padding(.horizontal, 6)
                .padding(.vertical, 3)
                .background(Chrome.accent, in: RoundedRectangle(cornerRadius: 5))
                .padding(7)
        }
    }

    /// Re-read the chart, or take it off the list. Lookout's own chart has
    /// neither: it is built from the sets below and cannot be removed.
    @ViewBuilder private var menu: some View {
        if onRefresh != nil || onRemove != nil {
            Menu {
                if let onRefresh {
                    Button("Read This Chart Again", systemImage: "arrow.clockwise", action: onRefresh)
                }
                if let onRemove {
                    Button("Remove", systemImage: "minus.circle", role: .destructive, action: onRemove)
                }
            } label: {
                Image(systemName: "ellipsis")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(Chrome.ink)
                    .frame(width: 22, height: 22)
                    .background(Chrome.surface.opacity(0.92), in: Circle())
                    .overlay(Circle().strokeBorder(Chrome.edge.opacity(0.6), lineWidth: 0.5))
            }
            .menuStyle(.borderlessButton)
            .menuIndicator(.hidden)
            .fixedSize()
            .padding(7)
            .accessibilityLabel("More for \(name)")
        }
    }
}


/// The last tile: add a chart by link or from a file.
private struct AddChartTile: View {
    let add: () -> Void

    var body: some View {
        Button(action: add) {
            VStack(spacing: 8) {
                Image(systemName: "plus")
                    .font(.system(size: 20, weight: .regular))
                    .foregroundStyle(Chrome.accent)
                Text("Add a chart")
                    .font(.system(size: 12.5, weight: .medium))
                    .foregroundStyle(Chrome.ink)
                Text("Style link, TileJSON, or a file")
                    .font(.system(size: 10.5))
                    .foregroundStyle(Chrome.muted)
                    .multilineTextAlignment(.center)
            }
            .padding(.horizontal, 12)
            .frame(width: ChartGallery.Metrics.add)
            .frame(maxHeight: .infinity)
            .background(Chrome.ink.opacity(0.02),
                        in: RoundedRectangle(cornerRadius: 11, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 11, style: .continuous)
                .strokeBorder(Chrome.edge.opacity(0.7),
                              style: StrokeStyle(lineWidth: 1.5, dash: [5, 4])))
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("add-chart-tile")
    }
}

File: android/app/src/main/java/org/beetlebug/lookout/charts/ChartGallery.kt
package org.beetlebug.lookout.charts

import org.beetlebug.lookout.R

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.outlined.Public
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

/**
 * The charts to draw, as tiles.
 *
 * One chart draws at a time. Lookout's own chart is built from the installed
 * sets; a link is a publisher's style drawn instead of it. A row of tiles
 * rather than a list, so two styles with similar names are told apart by
 * looking.
 *
 * A shipped style carries its picture in the app. One the mariner pasted shows
 * its kind and its url: rendering a style needs it resolved and its tiles
 * fetched, and this shell has no second engine to draw it with.
 */
@Composable
fun ChartGallery(
    links: ChartLinkController,
    /** How many cells the mariner's own chart is built from. */
    cells: Int,
    onAdd: () -> Unit,
) {
    Row(
        Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState())
            .padding(horizontal = 12.dp, vertical = 2.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        ChartTile(
            name = "Lookout chart",
            detail = if (cells > 0) "From your chart sets · $cells cells"
                     else "From your chart sets",
            active = links.activeChartLink == null,
            art = R.drawable.welcome_chart,
            onSelect = { links.selectChartLink(null) },
        )
        for (link in links.chartLinks) {
            ChartTile(
                name = link.name,
                detail = link.url,
                active = links.activeChartLink == link.url,
                art = ChartCatalog.art(link.url),
                onSelect = { links.selectChartLink(link.url) },
                onRefresh = { links.refreshChartLink(link.url) },
                onRemove = { links.removeChartLink(link.url) },
            )
        }
        AddChartTile(onAdd)
    }
}

private val TILE = 196.dp
private val ART = 112.dp
private val ADD = 150.dp

@Composable
private fun ChartTile(
    name: String,
    detail: String,
    active: Boolean,
    art: Int?,
    onSelect: () -> Unit,
    onRefresh: (() -> Unit)? = null,
    onRemove: (() -> Unit)? = null,
) {
    Surface(
        modifier = Modifier
            .width(TILE)
            .selectable(selected = active, role = Role.RadioButton, onClick = onSelect)
            .semantics { contentDescription = "chart-tile-$name" },
        shape = RoundedCornerShape(11.dp),
        color = MaterialTheme.colorScheme.surface,
        border = BorderStroke(
            if (active) 2.dp else 1.dp,
            if (active) MaterialTheme.colorScheme.primary
            else MaterialTheme.colorScheme.outlineVariant,
        ),
    ) {
        Column {
            Box(
                Modifier
                    .width(TILE)
                    .height(ART)
                    .clip(RoundedCornerShape(topStart = 10.dp, topEnd = 10.dp)),
            ) {
                if (art != null) {
                    Image(
                        painterResource(art),
                        contentDescription = null,
                        modifier = Modifier.fillMaxSize(),
                        contentScale = ContentScale.Crop,
                    )
                } else {
                    // A style with no picture shows its kind. It says the tile
                    // is a chart from somewhere else without claiming to be a
                    // portrayal of it.
                    Box(
                        Modifier.fillMaxSize().background(MaterialTheme.colorScheme.surfaceVariant),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(
                            Icons.Outlined.Public,
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.primary.copy(alpha = 0.5f),
                            modifier = Modifier.size(26.dp),
                        )
                    }
                }
                if (active) {
                    Text(
                        "ACTIVE",
                        style = MaterialTheme.typography.labelSmall,
                        fontWeight = FontWeight.Bold,
                        color = Color.White,
                        modifier = Modifier
                            .padding(6.dp)
                            .background(
                                MaterialTheme.colorScheme.primary,
                                RoundedCornerShape(4.dp),
                            )
                            .padding(horizontal = 5.dp, vertical = 1.dp),
                    )
                }
                if (onRemove != null) {
                    tileMenu(
                        onRefresh = onRefresh,
                        onRemove = onRemove,
                        modifier = Modifier.align(Alignment.TopEnd),
                        name = name,
                    )
                }
            }
            Column(Modifier.padding(horizontal = 11.dp).padding(top = 9.dp, bottom = 11.dp)) {
                Text(
                    name,
                    style = MaterialTheme.typography.labelLarge,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    detail,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

/** What can be done to a linked chart, where the tile is. */
@Composable
private fun tileMenu(
    onRefresh: (() -> Unit)?,
    onRemove: () -> Unit,
    modifier: Modifier,
    name: String,
) {
    var open by remember { mutableStateOf(false) }
    Box(modifier) {
        IconButton(onClick = { open = true }, modifier = Modifier.size(32.dp)) {
            Icon(
                Icons.Filled.MoreVert,
                contentDescription = "More for $name",
                tint = MaterialTheme.colorScheme.onSurface,
            )
        }
        DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            if (onRefresh != null) {
                DropdownMenuItem(
                    text = { Text("Read the style again") },
                    onClick = { open = false; onRefresh() },
                )
            }
            DropdownMenuItem(
                text = { Text("Remove") },
                onClick = { open = false; onRemove() },
            )
        }
    }
}

@Composable
private fun AddChartTile(onAdd: () -> Unit) {
    Surface(
        modifier = Modifier
            .width(ADD)
            .heightIn(min = ART + 60.dp)
            .clickable(onClick = onAdd)
            .semantics { contentDescription = "add-chart-tile" },
        shape = RoundedCornerShape(11.dp),
        color = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f),
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant),
    ) {
        Column(
            Modifier.fillMaxSize().padding(12.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            Icon(
                Icons.Filled.Add,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(24.dp),
            )
            Spacer(Modifier.height(8.dp))
            Text(
                "Add a chart",
                style = MaterialTheme.typography.labelLarge,
                fontWeight = FontWeight.Medium,
            )
            Text(
                "Style link or TileJSON",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}
